src/frontend/src/binder/expr/function/builtin_scalar.rs

```diff
@@ -434,6 +434,57 @@ impl Binder {
                 ("inner_product", raw_call(ExprType::InnerProduct)),
                 ("vector_norm", raw_call(ExprType::L2Norm)),
                 ("l2_normalize", raw_call(ExprType::L2Normalize)),
+                ("subvector", guard_by_len(|_, [vector_expr, start_expr, len_expr]| {
+                    let dimensions = if let DataType::Vector(length) = vector_expr.return_type() {
+                        length as i32
+                    } else {
+                        return Err(ErrorCode::BindError("subvector expects `vector(dim)` input".into()).into());
+                    };
+                    let start = start_expr
+                        .try_fold_const()
+                        .transpose()?
+                        .and_then(|datum| match datum {
+                            Some(ScalarImpl::Int32(v)) => Some(v),
+                            _ => None,
+                        })
+                        .ok_or_else(|| ErrorCode::ExprError("`start` must be an Int32 constant".into()))?;
+
+                    let len = len_expr
+                        .try_fold_const()
+                        .transpose()?
+                        .and_then(|datum| match datum {
+                            Some(ScalarImpl::Int32(v)) => Some(v),
+                            _ => None,
+                        })
+                        .ok_or_else(|| ErrorCode::ExprError("`count` must be an Int32 constant".into()))?;
+                    if len < 1 {
+                        return Err(ErrorCode::InvalidParameterValue(format!("vector must have at least 1 dimension, not {}", len)).into());
+                    }
+
+                    let end = start + len - 1;
+
+                    if start < 1 || end < 1 || start > dimensions || end > dimensions {
```

Comment:
Since we ensure that `len >= 1`, in most cases we can ensure that `start <= end`, and therefore we only need to check `start < 1 || end > dimension`. But we may have malicious unreasonably large input `len` to make the `end = start + len - 1` overflow to a minus value, which still satisfies `end < dimension`. 

So I think before this check, we can pre check the `len` to be `1 <= len <= MAX_VECTOR_SIZE`, and then the current check can be simplified to `start < 1 || end > dimension`.